x/subscription/keeper/subscription_test.go

```diff
@@ -2432,3 +2432,132 @@ func TestProjectsNumEnforcement(t *testing.T) {
+	require.Len(t, freePlan.AllowedBuyers, 0)
+	_, err := ts.TxSubscriptionBuy(consumer1, consumer1, freePlan.Index, 1, false, false)
+	require.NoError(t, err)
+
+	// change the plan's allowed buyers list - only consumer2 can purchase now
+	freePlan.AllowedBuyers = append(freePlan.AllowedBuyers, consumer2)
+	err = ts.TxProposalAddPlans(freePlan)
+	require.NoError(t, err)
+
+	// try making the subscription auto renew -> should fail
+	err = ts.TxSubscriptionAutoRenewal(consumer1, consumer1, freePlan.Index, true)
+	require.Error(t, err)
+
+	// buy subscription with consumer2 (with auto-renewal)
+	_, err = ts.TxSubscriptionBuy(consumer2, consumer2, freePlan.Index, 1, true, false)
+	require.NoError(t, err)
+
+	// change the plan's allowed buyers list - only consumer1 can purchase now
+	// note, we can't make it empty because empty list allows everyone to buy
+	freePlan.AllowedBuyers = []string{consumer1}
+	err = ts.TxProposalAddPlans(freePlan)
+	require.NoError(t, err)
+
+	// advance month to trigger the subscription's auto-renewal
+	ts.AdvanceMonths(1).AdvanceEpoch()
+
+	// the auto-renewal should have failed, so we're not supposed to find a subscription for consumer2
+	_, found := ts.getSubscription(consumer2)
+	require.False(t, found)
+}
+
+// TestAllowedBuyersFutureSubscription checks that when specifying a future subscription the creator
+// is allowed to buy it
+// Scenarios:
+// 1. buy a valid subscription with a future subscription that the user is not allowed to purchase -> should fail
+func TestAllowedBuyersFutureSubscription(t *testing.T) {
+	ts := newTester(t)
+	ts.SetupAccounts(2, 0, 0) // 2 sub, 0 adm, 0 dev
+	_, consumer1 := ts.Account("sub1")
+	_, consumer2 := ts.Account("sub2")
+
+	// buy subscription with empty allowed buyers list (without auto-renewal)
+	freePlan := ts.Plan("free")
+	require.Len(t, freePlan.AllowedBuyers, 0)
+	_, err := ts.TxSubscriptionBuy(consumer1, consumer1, freePlan.Index, 1, false, false)
+	require.NoError(t, err)
+
+	// change the premium plan's allowed buyers list - only consumer2 can purchase now
+	premiumPlan := ts.Plan("premium")
+	premiumPlan.AllowedBuyers = append(premiumPlan.AllowedBuyers, consumer2)
+	err = ts.TxProposalAddPlans(premiumPlan)
+	require.NoError(t, err)
+
+	// try buying a future subscription for the premium plan -> should fail
+	_, err = ts.TxSubscriptionBuy(consumer1, consumer1, premiumPlan.Index, 1, false, true)
+	require.Error(t, err)
+}
+
+// TestAllowedBuyersUpgradeSubscription checks that a user can't upgrade to a subscription it's not allowed to buy
+// Scenarios:
+// 1. try to upgrade subscription to a plan the user is not allowed to purchase -> should fail
+func TestAllowedBuyersUpgradeSubscription(t *testing.T) {
+	ts := newTester(t)
+	ts.SetupAccounts(2, 0, 0) // 2 sub, 0 adm, 0 dev
+	_, consumer1 := ts.Account("sub1")
+	_, consumer2 := ts.Account("sub2")
+
+	// buy subscription with empty allowed buyers list (without auto-renewal)
+	freePlan := ts.Plan("free")
```

Comment:
check that it works for cosumer2

Reply:
done. see https://github.com/lavanet/lava/commit/c35bb71ec4ea555c4e7ec9670cdc91007255c476